addons/mail/controllers/main.py

```diff
@@ -98,7 +100,11 @@ def _redirect_to_record(cls, model, res_id, access_token=None, **kwargs):
         if view_id:
             url_params['view_id'] = view_id
 
-        url = '/web?#%s' % url_encode(url_params)
+        if 'company_id' in record_sudo and record_sudo.company_id in user.company_ids:
+            cids = 'cids=%s' % (record_sudo.company_id.id)
+        else:
+            cids = ''
```

Comment:
- if there is no `company_id` on the record, use `cids=user.company_ids.ids + join`, otherwise it will crash for records with indirect rules, like leaves
- /!\ should be moved to hash, i.e. in url_params